src/cascadia/inc/cppwinrt_utils.h

```diff
@@ -55,4 +55,34 @@ private:
 // Use this if you have a Windows.Foundation.TypedEventHandler
 #define DEFINE_EVENT_WITH_TYPED_EVENT_HANDLER(className, name, eventHandler, sender, args)                                                        \
     winrt::event_token className::name(Windows::Foundation::TypedEventHandler<sender, args> const& handler) { return eventHandler.add(handler); } \
-    void className::name(winrt::event_token const& token) noexcept { eventHandler.remove(token); }
\ No newline at end of file
+    void className::name(winrt::event_token const& token) noexcept { eventHandler.remove(token); }
+
+// This is a helper method for deserializing a SAFEARRAY of
+// COM objects and converting it to a vector that
+// owns the extracted COM objects
+        template<typename T>
+        std::vector<wil::com_ptr<T>> SafeArrayToOwningVector(SAFEARRAY* safeArray)
+        {
+            T** pVals;
+            THROW_IF_FAILED(SafeArrayAccessData(safeArray, (void**)&pVals));
+
+            long lBound, uBound;
+            THROW_IF_FAILED(SafeArrayGetLBound(safeArray, 1, &lBound));
+            THROW_IF_FAILED(SafeArrayGetUBound(safeArray, 1, &uBound));
+
+            long count = uBound - lBound + 1;
+
+            std::vector<wil::com_ptr<T>> result;
+            result.reserve(count);
+
+            for (int i = 0; i < count; i++)
+            {
+                T* item = pVals[i];
+
+                wil::com_ptr<T> vec_elem;
+                vec_elem.attach(item);
+                result.emplace_back(vec_elem);

```

Comment:
This is one of the few times you'll want to pre-size instead of "reserve-and-emplace":

```
std::vector<wil::com_ptr<T>> result{count};
for(int i ...) {
    result[i].assign(pVals[i]);
}
```

since constructing the vector like this provides space for N pre-allocated and fully constructed com_ptrs, you can avoid the temporary by assigning directly into each com_ptr in turn.